### PR TITLE
fix(developer): warning fixes in kmcmplib 🙀

### PR DESCRIPTION
**Pull request by srl295**
- failing build on XCode
- fix sprintf, tmpnam, others

Fixes: #8688

Re-do of #8689 

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by srl295** [comment]
forgot the :gato_gritando:

File: developer/src/ext/json-schema-validator/json-validator.cpp
Comment (mcdurdin) [inline developer/src/ext/json-schema-validator/json-validator.cpp:1]: This is from an external repo. Have they got an updated version that we should be using instead?

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 1]
```diff
@@ -284,14 +284,17 @@ KMX_BOOL AddCompileError(KMX_DWORD msg)
     strcpy(szText, szTextp);
   }
   else {
-    sprintf(szText, "Unknown error %x", msg);
+    snprintf(szText, 793, "Unknown error %x", msg);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:287]: Let's fix it with proper constants:

```c++
#define COMPILE_ERROR_MAX_LEN (SZMAX_ERRORTEXT + 1 + 280)
```

Then, fix the def of `szText` at the top of the function, and:

```suggestion
    snprintf(szText, COMPILE_ERROR_MAX_LEN, "Unknown error %x", msg);
```

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 2]
```diff
@@ -284,14 +284,17 @@ KMX_BOOL AddCompileError(KMX_DWORD msg)
...
   }
 
-  if (kmcmp::ErrChr > 0)
-    sprintf(strchr(szText, 0), " character offset: %d", kmcmp::ErrChr);
+  if (kmcmp::ErrChr > 0) {
+    char *szTextNull = strchr(szText, 0);
+    snprintf(szTextNull, 793-(szTextNull-szText), " character offset: %d", kmcmp::ErrChr);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:292]:
```suggestion
    snprintf(szTextNull, COMPILE_ERROR_MAX_LEN-(szTextNull-szText), " character offset: %d", kmcmp::ErrChr);
```

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 3]
```diff
@@ -284,14 +284,17 @@ KMX_BOOL AddCompileError(KMX_DWORD msg)
...
+  }
 
   if (*ErrExtraLIB) {
-    sprintf(strchr(szText, 0), "%s", ErrExtraLIB);
+    char *szTextNull = strchr(szText, 0);
+    snprintf(szTextNull, 793-(szTextNull-szText), "%s", ErrExtraLIB);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:297]:
```suggestion
    snprintf(szTextNull, COMPILE_ERROR_MAX_LEN-(szTextNull-szText), "%s", ErrExtraLIB);
```

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 4]
```diff
@@ -1270,7 +1273,7 @@ KMX_BOOL CheckContextStatementPositions(PKMX_WCHAR context) {
 /**
  *  Checks if a use() statement is followed by other content in the output of a rule
  */
-KMX_DWORD CheckUseStatementsInOutput(const PFILE_GROUP gp,PKMX_WCHAR output) {
+KMX_DWORD CheckUseStatementsInOutput(const PFILE_GROUP /*gp*/,PKMX_WCHAR output) {
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:1276]: Let's fix the declaration and use if `gp` is not used:

```suggestion
KMX_DWORD CheckUseStatementsInOutput(PKMX_WCHAR output) {
```

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 5]
```diff
@@ -1731,7 +1734,7 @@ KMX_DWORD GetXString(PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX_WCHAR const * token,
 }
 
 KMX_DWORD GetXStringImpl(PKMX_WCHAR tstr, PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX_WCHAR const * token,
-  PKMX_WCHAR output, int max, int offset, PKMX_WCHAR *newp, int isVKey, int isUnicode
+  PKMX_WCHAR output, int max, int offset, PKMX_WCHAR *newp, int /*isVKey*/, int isUnicode
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:1737]:
```suggestion
  PKMX_WCHAR output, int max, int offset, PKMX_WCHAR *newp, int isUnicode
```

And fix callers...

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 6]
```diff
@@ -1794,7 +1797,7 @@ KMX_DWORD GetXStringImpl(PKMX_WCHAR tstr, PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX
     case 99:
       if (tokenFound) break;
       {
-        sprintf(ErrExtraLIB,"token: %c",(int)*p);
+        snprintf(ErrExtraLIB, 256, "token: %c",(int)*p);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:1800]: Shall we add a #define for these as well?

File: developer/src/kmcmplib/tests/api-test.cpp [hunk 7]
```diff
@@ -54,8 +54,11 @@ void setup() {
 
 void test_kmcmp_CompileKeyboardFile() {
   char kmn_file[L_tmpnam], kmx_file[L_tmpnam];
-  tmpnam(kmn_file);
-  tmpnam(kmx_file);
+  strncpy(kmn_file, "test_kmcmp_CompileKeyboardFile.kmn", L_tmpnam);
+  strncpy(kmx_file, "test_kmcmp_CompileKeyboardFile.kmx", L_tmpnam);
+
+  puts(kmn_file);
+  puts(kmx_file);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/api-test.cpp:61]: Is this because tmpnam is not available? Feels like a step backwards

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I'm still puzzled why you are running into these issues with kmcmplib. The build for kmc-kmn only uses build:wasm so you shouldn't be encountering these errors?

But changes are generally fine, although I am keen to try and improve code quality incrementally as we go

File: developer/src/kmcmplib/tests/api-test.cpp (hunk 7 above)
Comment (srl295) [reply developer/src/kmcmplib/tests/api-test.cpp:61]: It's marked as deprecated. And warnings-are-errors is on.

**Comment by srl295** [comment]
> I'm still puzzled why you are running into these issues with kmcmplib. The build for kmc-kmn only uses build:wasm so you shouldn't be encountering these errors?

OK. I was trying to build kmcmplib directly or from developer/build.sh which seemed to default to the local build.

In any event, even if it passed on wasm, I would have wanted this to be able to iterate and debug locally. 

> But changes are generally fine, although I am keen to try and improve code quality incrementally as we go

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 4 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp:1276]: +1.  Wasn't sure if it was meant to be a parallel to something else, etc.

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 6 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp:1800]: +1

Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: done

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 5 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: There's only one caller, however `GetXString()` has lots of callers (38, scattered) with both TRUE and FALSE values here.  Is this related to the commented out lines somehow?

```c
    case '[':  type = 11; break;	// start of vkey section
    //case ']':  type = 12; break;	// end of vkey section
    //case 'K':  type = 13; break;	// virtual key name or "key"
```

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 3 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: done

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 4 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: done

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 5 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: GetXStringImpl fixed, GetXString is not (above comment)

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 1 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: done

File: developer/src/ext/json-schema-validator/json-validator.cpp
Comment (srl295) [reply developer/src/ext/json-schema-validator/json-validator.cpp:1]: https://github.com/pboettch/json-schema-validator is newer, at a quick glance seems to not have the aliasing issue.

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 2 above)
Comment (srl295) [reply developer/src/kmcmplib/src/Compiler.cpp]: done

File: developer/src/kmcmplib/tests/api-test.cpp (hunk 7 above)
Comment (mcdurdin) [reply developer/src/kmcmplib/tests/api-test.cpp:61]: OK, we'll need to find a way to keep randomized temp file names for those situations where we have multiple processes using same temp

File: developer/src/ext/json-schema-validator/json-validator.cpp
Comment (mcdurdin) [reply developer/src/ext/json-schema-validator/json-validator.cpp:1]: Seems like a sensible update

**Comment by srl295** [comment]
> All good except for tmpnam which we need to find a solution for!

PTAL no longer changes tmpnam

**Comment by srl295** [comment]
I merged in #8696 here which reverts the tempnam and the json changes.

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.100-alpha](https://keyman.com/downloads/releases/alpha/17.0.100)

**Pull request merged**